python-checks/src/main/java/org/sonar/python/checks/GraphQLDenialOfServiceCheck.java

```diff
@@ -0,0 +1,146 @@
+
+  private static final List<String> VALID_MIDDLEWARE_NAMES = List.of("DEPTH", "COST");
+
+  private static final Predicate<String> VALID_MIDDLEWARE_PREDICATE = name ->
+    VALID_MIDDLEWARE_NAMES.stream().anyMatch(mwName -> name.toUpperCase(Locale.ROOT).contains(mwName));
+  private static final String MESSAGE_DEPTH = "Change this code to limit the depth of GraphQL queries.";
+  private static final String MESSAGE_CIRCULAR = "This relationship creates circular references.";
+
+  private final Set<Tree> circularReferences = new HashSet<>();
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FILE_INPUT, this::searchForCircularReferences);
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, this::checkGraphQLDepthLimit);
+  }
+
+  private void searchForCircularReferences(SubscriptionContext ctx) {
+    circularReferences.clear();
+    CircularRelationshipVisitor classVisitor = new CircularRelationshipVisitor(circularReferences);
+    ctx.syntaxNode().accept(classVisitor);
+  }
+
+  private void checkGraphQLDepthLimit(SubscriptionContext ctx) {
+    if (circularReferences.isEmpty()) {
+      // Query depth should only be limited if there are circular references in the database
+      return;
+    }
+    CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+    List<Argument> arguments = callExpression.arguments();
+
+    Optional.of(callExpression)
+      .map(CallExpression::callee)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(QualifiedExpression.class))
+      .filter(GraphQLUtils::isCallToAsView)
+      .map(QualifiedExpression::qualifier)
+      .filter(HasSymbol.class::isInstance)
+      .map(HasSymbol.class::cast)
+      .map(HasSymbol::symbol)
+      .filter(GraphQLUtils::isOrExtendsGraphQLView)
+      .filter(fqn -> !hasSafeValidationRules(arguments))
+      .filter(fqn -> !hasSafeMiddlewares(arguments))
+      .ifPresent(fqn -> {
+        PreciseIssue preciseIssue = ctx.addIssue(callExpression.callee(), MESSAGE_DEPTH);
+        circularReferences.forEach(circularRef -> preciseIssue.secondary(circularRef, MESSAGE_CIRCULAR));
+      });
+  }
+
+  private static boolean hasSafeMiddlewares(List<Argument> arguments) {
+    RegularArgument argument = TreeUtils.argumentByKeyword("middleware", arguments);
+    if (argument == null) {
+      return false;
+    }
+
+    return GraphQLUtils.extractListOrTupleArgumentValues(argument)
+      .map(values -> GraphQLUtils.expressionsNameMatchPredicate(values, VALID_MIDDLEWARE_PREDICATE))
+      .orElse(true);
+  }
+
+  private static boolean hasSafeValidationRules(List<Argument> arguments) {
+    RegularArgument argument = TreeUtils.argumentByKeyword("validation_rules", arguments);
+    if (argument == null) {
+      return false;
+    }
+
+    return GraphQLUtils.extractListOrTupleArgumentValues(argument)
+      .map(values -> (GraphQLUtils.expressionsNameMatchPredicate(values, VALID_MIDDLEWARE_PREDICATE)
```

Comment:
what if instead of having the predicate extract `GraphQLUtils.expressionsNameMatchPredicate(values, VALID_MIDDLEWARE_PREDICATE)` to method accepting just values? like
```java
private static boolean isValidMiddlewareNames(values) {
  return GraphQLUtils.expressionsNameMatchPredicate(values, name -> VALID_MIDDLEWARE_NAMES.contains(name.toUpperCase(Locale.ROOT))))
}
```
and then `VALID_MIDDLEWARE_PREDICATE` is not needed